pytorch_lightning/metrics/metric.py

```diff
@@ -0,0 +1,220 @@
+from abc import ABC, abstractmethod
+from typing import Any, Optional, Union
+
+import torch
+import torch.distributed
+
+from pytorch_lightning.metrics.convertors import tensor_metric, numpy_metric
+from pytorch_lightning.utilities.apply_to_collection import apply_to_collection
+
+__all__ = ['Metric', 'TensorMetric', 'NumpyMetric']
+
+
+class Metric(torch.nn.Module, ABC):
+    def __init__(self, name: str):
+        """
+        Abstract Base Class for metric implementation.
+        Should be used to implement metrics that
+        1.) Return multiple Outputs
+        2.) Handle their own DDP sync
+
+        Args:
+            name: the metric's name
+
+        """
+        super().__init__()
+        self.name = name
+        self._dtype = torch.get_default_dtype()
+        self._device = torch.device('cpu')
+
+    @property
+    def dtype(self) -> Union[str, torch.dtype]:
+        return self._dtype
+
+    @dtype.setter
+    def dtype(self, new_dtype: Union[str, torch.dtype]):
+        # Necessary to avoid infinite recursion
+        raise RuntimeError('Cannot set the dtype explicitly. Please use metric.to(new_dtype).')
+
+    @property
+    def device(self) -> Union[str, torch.device]:
+        return self._device
+
+    @device.setter
+    def device(self, new_device: Union[str, torch.device]):
+        # Necessary to avoid infinite recursion
+        raise RuntimeError('Cannot set the device explicitly. Please use metric.to(new_device).')
+
+    @abstractmethod
+    def forward(self, *args, **kwargs) -> torch.Tensor:
+        """
+        Implements the actual metric computation.
+
+        Returns:
+            metric value
+
+        """
+        raise NotImplementedError
+
+    def to(self, *args, **kwargs) -> torch.nn.Module:
+        """Moves and/or casts the parameters and buffers.
+
+        This can be called as
+
+        .. function:: to(device=None, dtype=None, non_blocking=False)
+
+        .. function:: to(dtype, non_blocking=False)
+
+        .. function:: to(tensor, non_blocking=False)
```

Comment:
could we have a doctest as an example instead?

Reply:
Done